openssl-sys/build.rs

```diff
@@ -1,86 +1,323 @@
+}
+
+/// Attempt to find OpenSSL through pkg-config.
+///
+/// Note that if this succeeds then the function does not return as pkg-config
+/// typically tells us all the information that we need.
+fn try_pkg_config() {
+    let target = env::var("TARGET").unwrap();
+    let host = env::var("HOST").unwrap();
+
+    // If we're going to windows-gnu we can use pkg-config, but only so long as
+    // we're coming from a windows host.
+    //
+    // Otherwise if we're going to windows we probably can't use pkg-config.
+    if target.contains("windows-gnu") && host.contains("windows") {
+        env::set_var("PKG_CONFIG_ALLOW_CROSS", "1");
+    } else if target.contains("windows") {
+        return
+    }
+
+    // We're going to be looking at header files, so show us all the system
+    // cflags dirs for showing us lots of `-I`.
+    env::set_var("PKG_CONFIG_ALLOW_SYSTEM_CFLAGS", "1");
+
+    let lib = match pkg_config::find_library("openssl") {
+        Ok(lib) => lib,
+        Err(_) => return,
+    };
+
+    if lib.include_paths.len() == 0 {
+        panic!("
+
+Used pkg-config to discover the OpenSSL installation, but pkg-config did not
+return any include paths for the installation. This crate needs to take a peek
+at the header files so it cannot proceed unless they're found.
+
+You can try fixing this by setting the `OPENSSL_DIR` environment variable
+pointing to your OpenSSL installation.
+
+");
+    }
+
+    validate_headers(&lib.include_paths, &lib.link_paths);
+
+    for include in lib.include_paths.iter() {
+        println!("cargo:include={}", include.display());
+    }
+
+    std::process::exit(0);
+}
+
+/// Validates the header files found in `include_dir` and then returns the
+/// version string of OpenSSL.
+fn validate_headers(include_dirs: &[PathBuf],
+                    libdirs: &[PathBuf]) -> String {
+    // This `*-sys` crate only works with OpenSSL 1.0.1, 1.0.2, and 1.1.0. To
+    // correctly expose the right API from this crate, take a look at
+    // `opensslv.h` to see what version OpenSSL claims to be.
+    let mut version_header = String::new();
+    let mut include = include_dirs.iter()
+                                  .map(|p| p.join("openssl/opensslv.h"))
+                                  .filter(|p| p.exists());
+    let mut f = match include.next() {
+        Some(f) => File::open(f).unwrap(),
+        None => {
+            panic!("failed to open header file at `openssl/opensslv.h` to learn
+                    about OpenSSL's version number, looked inside:\n\n{:#?}\n\n",
+                   include_dirs);
         }
     };
+    f.read_to_string(&mut version_header).unwrap();
 
-    let mode = if env::var_os("OPENSSL_STATIC").is_some() {
-    	"static"
-    } else {
-    	"dylib"
+    // Do a bit of string parsing to find `#define OPENSSL_VERSION_TEXT "..."`
+    let version_line = version_header.lines().find(|l| {
+        l.contains("OPENSSL_VERSION_TEXT") && !l.contains("fips")
```

Comment:
There are multiple lines, e.g. I have:

```
# define OPENSSL_VERSION_NUMBER  0x1010002fL
# ifdef OPENSSL_FIPS
#  define OPENSSL_VERSION_TEXT    "OpenSSL 1.1.0b-fips  26 Sep 2016"
# else
#  define OPENSSL_VERSION_TEXT    "OpenSSL 1.1.0b  26 Sep 2016"
# endif
```

But it might be easier to work with `OPENSSL_VERSION_NUMBER`.